File: 백준/Gold/2170. 선 긋기/선 긋기.py
import sys
import heapq
input = sys.stdin.readline

# 12:20
# 3차원 그래프
# 겹치는 구역 찾기 문제
# 모든 점 루프도는 부르트포스는 안되겠고
# n개 돌면서 2개 겹치는 부분 찾고
# 12:32 알고리즘 분류 확인 
# 3차원 누적합? 스위핑? 좌표 압축?
# n, k = map(int, input().split())
# fish_territory = [list(map(int, input().split())) for _ in range(n)]
# fish_territory.sort()
# prefix_sum = []


n = int(input())
lines = [tuple(map(int,input().split())) for _ in range(n)]

lines.sort()
answer = 0
s=lines[0][0]
e=lines[0][1]
for i in range(1,n):
    if e >= lines[i][0]:
        if e>=lines[i][1]:
            continue
        else:
            e = lines[i][1]
    else:
        answer+=(e-s)
        s = lines[i][0]
        e = lines[i][1]
answer+=(e-s)
print(answer)

File: 백준/Gold/2170. 선 긋기/선 긋기.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        // 빠른 입력
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());

        // 입력 받을 리스트
        List<int[]> lines = new ArrayList<>();

        for (int i = 0; i < n; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            int start = Integer.parseInt(st.nextToken());
            int end = Integer.parseInt(st.nextToken());
            lines.add(new int[]{start, end});
        }

        // 시작점을 기준으로 정렬
        lines.sort(Comparator.comparingInt(a -> a[0]));

        int answer = 0;
        int s = lines.get(0)[0];
        int e = lines.get(0)[1];

        for (int i = 1; i < n; i++) {
            int curStart = lines.get(i)[0];
            int curEnd = lines.get(i)[1];

            if (e >= curStart) {
                e = Math.max(e, curEnd);
            } else {
                answer += e - s;
                s = curStart;
                e = curEnd;
            }
        }
        answer += e - s;

        System.out.println(answer);
    }
}
